File: 2025/139. Word Break/Solution.java
class TrieNode {
    boolean isWord;
    Map<Character, TrieNode> map;
    public TrieNode() {
        isWord = false;
        map = new HashMap<>();
    }
}
class Solution {

    TrieNode trieRoot;

    public boolean wordBreak(String s, List<String> wordDict) {
        trieRoot = new TrieNode();
        buildTrie(wordDict);
        boolean[] isWord = new boolean[s.length()]; // memorize possible words locations, start from a word for next iteration
        for (int i = 0; i < s.length(); i++) {
            if (i == 0 || isWord[i - 1]) {
                TrieNode current = trieRoot;
                for (int j = i; j < s.length(); j++) {
                    char c = s.charAt(j);
                    if (!current.map.containsKey(c))
                        break;
                    current = current.map.get(c);
                    if(current.isWord)
                        isWord[j] = current.isWord;
                }
            }
        }
        return isWord[s.length() - 1];
    }

    private void buildTrie(List<String> words) {
        for (String s : words) {
            TrieNode current = trieRoot;
            for (char c : s.toCharArray()) {
                TrieNode next = current.map.get(c);
                if (next == null) {
                    next = new TrieNode();
                    current.map.put(c, next);
                }
                current = next;
            }
            current.isWord = true;
        }
    }
}

File: 2025/139. Word Break/solution.py
class TrieNode:
    def __init__(self):
        self.children = defaultdict(TrieNode)
        self.isWord = False

class Trie:
    def __init__(self):
        self.root = TrieNode()

    def insert(self, word: str):
        node = self.root
        for c in word:
            if c not in node.children:
                node.children[c] = TrieNode()
            node = node.children[c]
        node.isWord = True

    def search(self, word: str) -> bool:
        node = self.root
        for c in word:
            if c not in node.children:
                return False
            node = node.children[c]
        return node.isWord

    def buildTrie(self, wordDict: List[str]) -> TrieNode:
        for word in wordDict:
            self.insert(word)
        return self.root

class Solution:
    def wordBreak(self, s: str, wordDict: List[str]) -> bool:
        root = Trie()
        root.buildTrie(wordDict)
        length = len(s)
        is_word = [False] * length
        for i in range(len(s)):
            if i == 0 or is_word[i - 1]:
                for j in range(i, length):
                    if root.search(s[i:j+1]):
                        is_word[j] = True
        return is_word[-1]